core/src/test/java/apoc/util/UtilsTest.java

```diff
@@ -291,6 +302,129 @@ public void testMd5() {
+                        """
+                                .formatted(hashFunc, "p" + i, hashFunc, hashFunc);
+                final var params = Map.of("value", value, "list", List.of(value));
+                assertStableHash(seed, value, tx, query, params);
+            }
+
+            final var mapQuery =
+                    """
+                    match (n:HashFunctionsAreStable)
+                    return
+                      apoc.util.%s([properties(n)]) as hash1,
+                      apoc.util.%s($value) as hash2
+                    """
+                            .formatted(hashFunc, hashFunc);
+            assertStableHash(seed, randMap, tx, mapQuery, Map.of("value", List.of(randMap)));
+
+            final var listCypher = IntStream.range(0, randStorables.size())
+                    .mapToObj(i -> "n.p" + i)
+                    .collect(Collectors.joining(","));
+            final var listQuery =
+                    """
+                match (n:HashFunctionsAreStable)
+                return
+                  apoc.util.%s([%s]) as hash1,
+                  apoc.util.%s($value) as hash2
+                """
+                            .formatted(hashFunc, listCypher, hashFunc);
+            assertStableHash(seed, randStorables, tx, listQuery, Map.of("value", randStorables));
+
+            final var listOfListsQuery =
+                    """
+                match (n:HashFunctionsAreStable)
+                return
+                  apoc.util.%s([[%s]]) as hash1,
+                  apoc.util.%s([$value]) as hash2
+                """
+                            .formatted(hashFunc, listCypher, hashFunc);
+            assertStableHash(seed, randStorables, tx, listOfListsQuery, Map.of("value", randStorables));
+
+            final var nodesHashQuery =
+                """
+                match (a:HashFunctionsAreStable)-[r]->(b:HashFunctionsAreStable)
+                return
+                  apoc.util.%s([a, b, r, [a, b, r]]) as hash1,
+                  apoc.util.%s([a, b, r, [a, b, r]]) as hash2
+                """.formatted(hashFunc, hashFunc);
+            assertStableHash(seed, "nodes and rels", tx, nodesHashQuery, Map.of());
+        } finally {
+            db.executeTransactionally("cypher runtime=slotted match (n) detach delete n");
+        }
+    }
+
+    private static void assertStableHash(
+            long seed, Object val, Transaction tx, String query, Map<String, Object> params) {
+        final var a = tx.execute(query, params).stream().toList();
+        final var b = tx.execute(query, params).stream().toList();
+        final var message = "%s should have stable hash (seed=%s)".formatted(val, seed);
+        assertEquals(message, a, b);
+        assertEquals(1, a.size());
+        final var first = a.get(0).entrySet().iterator().next();
+        for (final var e : a.get(0).entrySet()) {
+            assertEquals(message, first.getValue(), e.getValue());
+        }
```

Comment:
Can you explain the point of this assertion? To me it looks like you are comparing something to itself

Reply:
I can, it's a bit messy. Let's look at an example:

We might run this query with `$value=23, $list=List(23)` and `n.p1=23`:

```
match (n)
return
  apoc.util.md5([n.p1]) as hash1,
  apoc.util.md5([$value]) as hash2
  apoc.util.md5($list) as hash3
```

This assertion is intended to make sure `hash1`, `hash2` and `hash3` have the same value. They should all equal each other, but in Cypher there can be subtle differences depending on where a value comes from. For example `create (n {p: [1,2,3]}) return n.p` will most probably not have exactly the same types internally as `return [1,2,3]` even though the two lists are equal and represents the same thing. Just wanted to make sure that such internal types do not affect the hash.

Reply:
I see, thanks for explaining!